okhttp/src/main/java/okhttp3/internal/tls/BasicCertificateChainCleaner.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.tls
+
+import java.security.GeneralSecurityException
+import java.security.cert.Certificate
+import java.security.cert.X509Certificate
+import java.util.Deque
+import java.util.ArrayDeque
+import javax.net.ssl.SSLPeerUnverifiedException
+
+/**
+ * A certificate chain cleaner that uses a set of trusted root certificates to build the trusted
+ * chain. This class duplicates the clean chain building performed during the TLS handshake. We
+ * prefer other mechanisms where they exist, such as with
+ * [okhttp3.internal.platform.AndroidPlatform.AndroidCertificateChainCleaner].
+ *
+ * This class includes code from <a href="https://conscrypt.org/">Conscrypt's</a> [TrustManagerImpl]
+ * and [TrustedCertificateIndex].
+ */
+class BasicCertificateChainCleaner(
+  private val trustRootIndex: TrustRootIndex
+): CertificateChainCleaner() {
+
+  /**
+   * Returns a cleaned chain for [chain].
+   *
+   * This method throws if the complete chain to a trusted CA certificate cannot be constructed.
+   * This is unexpected unless the trust root index in this class has a different trust manager than
+   * what was used to establish [chain].
+   */
+  @Throws(SSLPeerUnverifiedException::class)
+  override fun clean(chain: MutableList<Certificate>, hostname: String): MutableList<Certificate> {
```

Comment:
Parameter and return should be `List`.

Reply:
db2e652e414d93a134691ab80ffd4ccaadd06d7e